File: internal/permissions/permissions.go
package permissions

import (
	"encoding/json"
	"os"

	"github.com/samber/lo"
)

type Claims = map[string]string
type Permissions map[string][]Claims
type PermissionsList map[string]struct{}

func LoadFromJSONFile(filename string) (Permissions, error) {
	content, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	var data Permissions
	if err := json.Unmarshal(content, &data); err != nil {
		return nil, err
	}

	return data, nil
}

func (perms Permissions) GetPermissions(providedClaims Claims) PermissionsList {
	res := make(PermissionsList)
	for permission, requiredClaimsSlice := range perms {
		if lo.SomeBy(requiredClaimsSlice, func(requiredClaims Claims) bool {
			for k, v := range requiredClaims {
				if providedClaims[k] != v {
					return false
				}
			}
			return true
		}) {
			res[permission] = struct{}{}
		}
	}
	return res
}

func (perms PermissionsList) Has(permissions ...string) bool {
	return lo.EveryBy(permissions, func(permission string) bool {
		_, ok := perms[permission]
		return ok
	})
}

File: internal/permissions.ts
import {readFileSync} from "fs";

export type Claims = Record<string, string>;
export type Permissions = Record<string, Claims[]>;
export type PermissionsList = Set<string>;

export function LoadFromJSONFile(filename: string): Permissions {
  const content = readFileSync(filename);
  return JSON.parse(content.toString());
}

export function GetPermissions(perms: Permissions, providedClaims: Claims): PermissionsList {
  return new Set(...Object
    .entries(perms)
    .filter(([_, requiredClaims]) => requiredClaims.some(requiredClaims => Object
      .entries(requiredClaims)
      .every(([k, v]) => providedClaims[k] != v)))
    .map(([permission, _]) => permission));
}

export function Has(perms: PermissionsList, ...permissions: string[]): boolean {
  return permissions.every(permission => {
    return perms.has(permission);
  });
}
